[tcat] Add tcat implementations and bug fix.

arnulfrupp opened the pull request:
The PR is derived from the stale PR "[tcat] Add tcat implementaions and bug fix #11073".
Comments  provided for the old PR are addressed in this PR.

Scope:
- Commit adds check is commissioning is possible if tcat device is already commissioned.
- Adds advertisement update on disconnected and role change.
- Fix key handling for key references.
- Fixes the authorisation processing scheme.
- Implements recent changes of the applications TLVs

maciejbaczmanski commented on src/core/meshcop/tcat_agent.cpp:
```diff
@@ -187,99 +190,102 @@ void TcatAgent::Disconnected(void)
-        {
-            aDataset->ConvertTo(datasetInfo);
-        }
+            case kThreadDomainFlag:
```
`GetDomainName` is guarded with
```
#if (OPENTHREAD_CONFIG_THREAD_VERSION >= OT_THREAD_VERSION_1_2)
```
I think that this block should be guarded as well to avoid build failures?

arnulfrupp replied:
@maciejbaczmanski, thanks for approving and for the hint.

maciejbaczmanski reviewed (approved):
Thanks! One comment regarding failing CI

EskoDijk commented:
A bit of delay due to the MM - but I'll come to this review soon!

EskoDijk commented on tools/tcat_ble_client/tlv/tcat_tlv.py:
```diff
@@ -47,7 +47,14 @@ class TcatTLVType(Enum):
     DECOMMISSION = 0x60
-    APPLICATION = 0x82
+    GET_APPLICATION_LAYERS = 0x80
+    APPLICATION_1 = 0x81
```
For all these TLV names here, we should base it on the actual spec TLV names as much as possible. So e.g.

```suggestion
    APPLICATION_DATA_1 = 0x81
```

arnulfrupp replied:
fixed.

EskoDijk reviewed (changes requested):
@arnulfrupp Thanks! I provided some comments to consider, and a few fixes. We could also discuss this in a WG call.

arnulfrupp commented on include/openthread/ble_secure.h:
```diff
@@ -406,15 +406,20 @@ otError otBleSecureSend(otInstance *aInstance, uint8_t *aBuf, uint16_t aLength);
+ * @retval OT_ERROR_NONE             Successfully sent data.
+ * @retval OT_ERROR_NO_BUFS          Failed to allocate buffer memory.
+ * @retval OT_ERROR_INVALID_STATE    TLS connection was not initialized.
+ * @retval OT_ERROR_REJECTED         Application protocol is response with data or status but no response is pending.
```
Good point. The function description should be clearer now.

arnulfrupp commented on src/core/radio/ble_secure.hpp:
```diff
@@ -276,6 +281,14 @@ class BleSecure : public InstanceLocator, public MeshCoP::Tls::Extension, privat
     bool GetInstallCodeVerifyStatus(void) const { return mTcatAgent.GetInstallCodeVerifyStatus(); }
 
+    /**
+     * @brief Notifies BLE layer that advetrisement should be updated.
```
fixed

arnulfrupp commented on src/posix/platform/ble.cpp:
```diff
@@ -102,6 +102,14 @@ otError otPlatBleGapAdvSetData(otInstance *aInstance, uint8_t *aAdvertisementDat
     return OT_ERROR_NONE;
 }
 
+otError otPlatBleGapAdvUpdateData(otInstance *aInstance, uint8_t *aAdvertisementData, uint16_t aAdvertisementLen)
```
The POSIX platform is uses for simulation. Here BLE advertisement is not relevant.
Yes, the BR repo could implement TCAT advertisement. I believe for the core openthread repo where POSIX is only used for simulation it should be fine as is.

EskoDijk replied:
Agreed, OTBR could separately add this if needed. Also for BR management, TCAT over Wi-Fi or Ethernet would be possible also! In the same way that now a Commissioner can connect over DTLS to a BR.

arnulfrupp commented on tools/tcat_ble_client/cli/cli.py:
```diff
@@ -49,6 +50,12 @@ def __init__(self,
             'help': HelpCommand(),
             'hello': HelloCommand(),
+            'get_applications': GetApplicationLayersCommand(),
+            'application1': SendApplicationData1(),
```
fixed.

arnulfrupp commented on tools/tcat_ble_client/cli/cli.py:
```diff
@@ -49,6 +50,12 @@ def __init__(self,
+            'application2': SendApplicationData2(),
+            'application3': SendApplicationData3(),
+            'application4': SendApplicationData4(),
+            'vendor_application': SendVendorApplicationData(),
```
fixed.

arnulfrupp commented on tools/tcat_ble_client/cli/base_commands.py:
```diff
@@ -107,7 +107,102 @@ def get_help_string(self) -> str:
+                elif (tlv_application.type == TcatTLVType.SERVICE_NAME_TCP.value):
+                    print(f"\tApplication {i} is TCP service: {tlv_application.value.decode('ascii')}")
+                else:
+                    print('\tUnkonwn service type.')
```
fixed.

arnulfrupp commented on tools/tcat_ble_client/pyproject.toml:
```diff
@@ -12,5 +12,6 @@ authors = ["Piotr Jasinski <piotr.jasinski@nordicsemi.no>"]
 python = "^3.7"
 bleak = "^0.20.2"
 pytest ="^7.1.2"
+cryptography = ">=3.7"
```
fixed.

arnulfrupp commented:
@EskoDijk thank you for your review. All your comments have been addressed. Can you please check again.

arnulfrupp commented:
@EskoDijk some checks are now failing because nt-ns does not implement otPlatBleGapAdvUpdateData(..).
I see otPlatBleGapAdvSetData(...) just returns OT_ERROR_NOT_IMPLEMENTED and believe otPlatBleGapAdvUpdateData(..) could just do the same. Do you want me make a PR or may I leave this with you!?

EskoDijk commented:
> Yes, now with an all-new OTNS2! I expected some issue to pop up because of the change. I can submit a PR for it, sure.

So actually there wasn't an issue with OTNS2 - this pops up because the OTNS2 platform code doesn't know about new platform functions, obviously.
But I added a PR for it; which can be merged before approving this (to get CI tests green) or after this (when the Dependabot would indicate that bumping 'openthread' fails on OTNS2). 

See: https://github.com/openthread/ot-ns/pull/569

EskoDijk commented on include/openthread/tcat.h:
```diff
@@ -156,27 +166,30 @@ typedef struct otTcatVendorInfo
 
 /**
- * Pointer to call when application data was received over a TCAT TLS connection.
+ * Pointer to call when vendor specific data was received over a TCAT TLS connection.
```
The data type name has "Application data" in it. As I understand it supports both the application-layer data, as well as the vendor-specific data (which is really just like another app-layer except it doesn't have a formal name defined).

So in keeping with the rest of the naming, maybe "application data" would be better? Or we say:

```suggestion
 * Pointer to call when application data or vendor-specific data was received over a TCAT TLS connection.
```

arnulfrupp replied:
Fixed.
I also copied some additional text from otHandleBleSecureReceive(...) here to describe that TCAT agent will respond with OT_TCAT_STATUS_UNSUPPORTED if no response is created inside the callback.

EskoDijk commented on src/cli/README_TCAT.md:
```diff
@@ -63,6 +63,25 @@ tcat advid clear
 
+### certid
+
+Displays ID of currently chosen device certificate.
```
```suggestion
Displays ID of currently selected TCAT Device certificate. A TCAT Device supports multiple identities for testing purposes.
```

arnulfrupp replied:
fixed

EskoDijk commented on src/cli/cli_tcat.cpp:
```diff
@@ -86,15 +117,18 @@
+const char kApplicationServiceName1[] = "echo";
+const char kApplicationServiceName2[] = "discard";
+
+uint8_t selectedCert = 0;
```
If a static var, is should be `sSelectedCert` I think?  Why can't it be a private member var of the `ot.Cli.Tcat` class? Next to `otTcatVendorInfo mVendorInfo;`.

arnulfrupp replied:
fixed

EskoDijk commented on src/core/meshcop/tcat_agent.cpp:
```diff
@@ -903,19 +896,110 @@ void TcatAgent::CalculateHash(uint64_t aChallenge, const char *aBuf, size_t aBuf
+    return error;
+}
+
+Error TcatAgent::HandlerApplicationData(const Message          &aIncomingMessage,
```
Was "Handler" in the name intended here, or should it be "Handle" as the other similar methods are named?

arnulfrupp replied:
fixed

EskoDijk commented on src/core/meshcop/tcat_agent.cpp:
```diff
@@ -903,19 +896,110 @@ void TcatAgent::CalculateHash(uint64_t aChallenge, const char *aBuf, size_t aBuf
+    mAppDataReceiveCallback.InvokeIfSet(&GetInstance(), &aIncomingMessage, aOffset,
+                                        static_cast<otTcatApplicationProtocol>(aApplicationProtocol));
+
+    if (mApplicationResponsePending)
```
Does this handle all cases correctly? E.g. if the sending of the response failed at the BLE layer, this 'pending' will still be true, even though the functionality is implemented and supported. (Or maybe I misunderstand the flow.)

arnulfrupp replied:
This present implementation just ensures that one TLV is provided as a response. This is important to not break the command/response message flow. If a response TLV (status or payload) has been successfully generated inside the application handler, the transport  layer (BleSecure) has the responsibility to call NotifyApplicationResponseSent(...).

You are right, the implementation does not differentiate between the case that no response has been generated because the application is not implemented, not properly implemented or unsuccessful. I suggest it is in the responsibility of the application layer to create the proper response also in case of special circumstances, e.g. it could respond with status OT_TCAT_STATUS_GENERAL_ERROR when sending a longer response failed earlier. If sending such a short reply would fail, the TCAT agent will probably also fail with his "not implemented" response. 

The Thread specification does not disclose further details which status codes shall be used here. The commissioner can always check if the TCAT device intended to implement a specific application layer using "Get Application Layers" TLV.

EskoDijk replied:
Thanks, I checked the flow again, and if the app-specific handler hits an error while creating the response message, before `mApplicationResponsePending` could be cleared, then the TcatAgent will subsequently always clear it in `HandleApplicationData`. Then, in `HandleSingleTlv` a status response with single error status value will be selected instead, and appended to the outgoing message, which should work because it's short. So that part looks ok. (I've found one unhappy flow in existing code, which I'll submit a separate PR for.)

EskoDijk commented on src/core/meshcop/tcat_agent.hpp:
```diff
@@ -183,10 +184,14 @@ class TcatAgent : public InstanceLocator, private NonCopyable
-        kTlvSelectApplicationLayerTcp = 129, ///< TCAT select TCP protocol application layer request TLV
-        kTlvSendApplicationData       = 130, ///< TCAT send application data TLV
-        kTlvSendVendorSpecificData    = 159, ///< TCAT send vendor specific command or data TLV
+        kTlvGetApplicationLayers   = 128, ///< TCAT get application layers request TLV
```
For all these I would prefer the hex numbers e.g. 0x80. This matches the spec better. Which makes it easier findable with Ctrl+F and in IDEs and webbrowsers, in both directions.

arnulfrupp replied:
fixed

EskoDijk reviewed (changes requested):
Reviewed again, added a few questions / suggestions!

arnulfrupp commented:
> Reviewed again, added a few questions / suggestions!

@EskoDijk thanks for reviewing again. You comments are addressed. Let me know of you agree with my explanation about pending response error handling in HandleApplicationData(...).
I also found and fixed an issue in CheckCommandClassAuthorizationFlags(..).  For set active operational dataset TLV the PSKc check must always be successful. 
Please review again.

arnulfrupp commented:
> > Would be great to merge the OTNS2 PR first. I'm not sure if this will solve all issues or if there is anything else.
> 
> Still, that is a bit of a strange PR - it adds an implementation for a method that's not defined (yet) in the header. Like a catch-22: both PRs need each other merged first.

I suggest we finish reviews for both PRs and later ask Jonathan how to resolve the PR deadlock. One of the PRs needs to run the red light.

EskoDijk commented:
@arnulfrupp Of all the failing tests, the one on "API Version" is what you could probably fix now.  The rest seems out of our control ? The OTNS ones have a PR waiting in the otns project.

abtink commented on include/openthread/tcat.h:
```diff
@@ -82,8 +83,10 @@ typedef enum otTcatStatusCode
-    OT_TCAT_STATUS_HASH_ERROR   = 7, ///< The hash value presented by the commissioner was incorrect
+    OT_TCAT_STATUS_UNDEFINED  = 6, ///< The requested value, data or service is not defined (currently) or not present
+    OT_TCAT_STATUS_HASH_ERROR = 7, ///< The hash value presented by the commissioner was incorrect
+    OT_TCAT_STATUS_INVALID_STATE = 8, /** The Command cannot be executed because the TCAT
```
Use `///<` instead of `/**`. or (can move the `/**` before the definition)?

arnulfrupp replied:
fixed

abtink commented on src/core/meshcop/tcat_agent.cpp:
```diff
@@ -188,100 +190,114 @@ void TcatAgent::Disconnected(void)
+#if (OPENTHREAD_CONFIG_THREAD_VERSION >= OT_THREAD_VERSION_1_4)
+                    if (Get<MeshCoP::NetworkNameManager>().GetDomainName() == mCommissionerDomainName)
+#else
+                    if (strcmp(mCommissionerDomainName.GetAsCString(), NetworkName::kDomainNameInit) == 0)
```
Using `strcmp()` is fine. Can also consider using `StringMatch()` here (returns a `bool` so easier to check for equality. Also we can use `StringMatchMode` to either do an exact match, or `kStringCaseInsensitiveMatch`).

arnulfrupp replied:
fixed

abtink commented on src/core/meshcop/tcat_agent.cpp:
```diff
@@ -903,19 +902,110 @@ void TcatAgent::CalculateHash(uint64_t aChallenge, const char *aBuf, size_t aBuf
+        replyLen += sizeof(tlv);
+        replyLen += StringLength(mVendorInfo->mApplicationServiceName[i], kServiceNameMaxLength);
+        count++;
+    }
```
This seems to calculate the TLV value length. 
Should we check that it does not overflow (goes above 255)?

If it can go above 255, should we then use an `ExtendedTlv` for `kTlvResponseWithPayload`?

arnulfrupp replied:
The TLV can contain maximum 4 sub TLVs with services names not exceeding 15 characters. Those values are fixed in specification. For that reason I suggest a normal TLV is safe.

abtink commented on src/core/meshcop/tcat_agent.cpp:
```diff
@@ -930,6 +1020,22 @@ Error TcatAgent::HandleStartThreadInterface(void)
+{
+    Error error;
+
+    if (!IsCommandClassAuthorized(kCommissioning))
+    {
+        error = kErrorRejected;
+        ExitNow();
+    }
```
How about?
```suggestion
    VerifyOrExit(IsCommandClassAuthorized(kCommissioning), error = kErrorRejected);
```

arnulfrupp replied:
fixed (in 7 places)

abtink reviewed (commented):
Looks good overall. Thanks @arnulfrupp.
Just added couple of smaller suggestions/questions below:

arnulfrupp commented:
@abtink your comment have been addressed. Please review again.

arnulfrupp commented:
> @abtink your comments have been addressed. Please review again.

@abtink kind reminder to review again.

abtink reviewed (approved):
LGTM.

arnulfrupp commented:
@jwhui this PR has been approved. Can we merge it?
Test /home/runner/work/openthread/openthread/tests/scripts/thread-cert/test_detach.py fails since a more recent rebases. I don't believe it is TCAT related. Let me me know if you think different.

jwhui commented:
@gemini-code-assist review

arnulfrupp commented on tools/tcat_ble_client/dataset/dataset_entries.py:
```diff
@@ -66,13 +66,13 @@ def __init__(self):
         self.ubit = value & 0x1
         self.ticks = (value >> 1) & 0x7FFF
-        self.seconds = (value >> 16) & 0xFFFF
+        self.seconds = (value >> 16) & 0xFFFFFFFFFFFF
```
fixed

arnulfrupp commented on src/core/meshcop/tcat_agent.cpp:
```diff
@@ -188,100 +190,114 @@ void TcatAgent::Disconnected(void)
-                extendedPanIdsMatch = true;
+            default:
+                LogCrit("Error while processing access flags. Unexpected flag %d", flag);
+                OT_ASSERT(false); // Should not get here
```
There is a insightful comment in the logging statement directly above. I suggest duplication of the comment in OT_ASSERT is not needed and the extra burden on flash consumption can be avoided.

Vyrastas commented on include/openthread/ble_secure.h:
```diff
@@ -409,17 +412,28 @@ otError otBleSecureSendMessage(otInstance *aInstance, otMessage *aMessage);
- * @retval OT_ERROR_NONE           Successfully sent data.
- * @retval OT_ERROR_NO_BUFS        Failed to allocate buffer memory.
- * @retval OT_ERROR_INVALID_STATE  TLS connection was not initialized.
+ * Sends a secure BLE data packet containing application data directed to the application layer @p  aApplicationProtocol
+ * or a response to the latest received application data packet.
+ * Only a single response can be sent while executing the otHandleBleSecureReceive handler. If no (further) response is
+ * expected OT_ERROR_REJECTED is returned.
+ * For response with payload @p aApplicationProtocol shall be set to OT_TCAT_APPLICATION_PROTOCOL_PAYLOAD.
+ * For response with status @p aApplicationProtocol shall be OT_TCAT_APPLICATION_PROTOCOL_STATUS and @ aBuf shall
+ * contain a single byte otTcatStatusCode value.
```
```suggestion
 * Sends a secure BLE data packet containing application data directed to the application layer @p  aApplicationProtocol
 * or a response to the latest received application data packet.
 *
 * Only a single response can be sent while executing the `otHandleBleSecureReceive` handler. If no (further) response is
 * expected `OT_ERROR_REJECTED` is returned.
 *
 * For responses with a payload @p aApplicationProtocol shall be set to `OT_TCAT_APPLICATION_PROTOCOL_PAYLOAD`.
 * For responses with a status @p aApplicationProtocol shall be `OT_TCAT_APPLICATION_PROTOCOL_STATUS` and @ aBuf shall
 * contain a single byte `otTcatStatusCode` value.
```

Vyrastas commented on include/openthread/tcat.h:
```diff
@@ -159,27 +169,33 @@ typedef struct otTcatVendorInfo
     const otTcatAdvertisedDeviceId *mAdvertisedDeviceIds; /** Vendor managed advertised device ID array.
-                                         Array is terminated like C string with OT_TCAT_DEVICE_ID_EMPTY */
+                                        Array is terminated like C string with OT_TCAT_DEVICE_ID_EMPTY */
     const otTcatGeneralDeviceId *mGeneralDeviceId;        /** Vendor managed general device ID array.
-                                               (if NULL: device ID is set to EUI-64 in binary format)*/
+                                        (if NULL: device ID is set to EUI-64 in binary format) */
+    const char *mApplicationServiceName[OT_TCAT_APPLICATION_LAYER_MAX_COUNT]; /** Array with
+                   application service names as C string with maximum length
+                   OT_TCAT_SERVICE_NAME_MAX_LENGTH or NULL if not supported */
+    bool mApplicationServiceIsTcp[OT_TCAT_APPLICATION_LAYER_MAX_COUNT];       /** Array with
+                         boolean values indicating if the service is of TCP type (otherwise UDP) */
 
```
Some of the comment formatting on this file seems odd, was make-pretty run?

Vyrastas commented on src/cli/README_TCAT.md:
```diff
@@ -63,6 +63,25 @@ tcat advid clear
 
+### certid
+
+Displays ID of currently selected TCAT Device certificate. A TCAT Device supports multiple identities for testing purposes.
```
```suggestion
Displays the ID of currently selected TCAT device certificate. A TCAT device supports multiple identities for testing purposes.
```

Vyrastas commented on src/cli/README_TCAT.md:
```diff
@@ -63,6 +63,25 @@ tcat advid clear
+
+### certid \<id\>
+
+Selects the ID of the TCAT Device certificate. A TCAT Device supports multiple identities for testing purposes.
```
```suggestion
Selects the ID of the TCAT device certificate. A TCAT device supports multiple identities for testing purposes.
```

Vyrastas commented on src/cli/cli_tcat.cpp:
```diff
@@ -230,6 +258,38 @@ template <> otError Tcat::Process<Cmd("advid")>(Arg aArgs[])
+ * @cparam tcat certid [@ca{value}]
+ * * The `value` int value of the ID.
+ * @par
+ * Selects predefined certificate.
```
```suggestion
 * Selects a predefined certificate.
```

Vyrastas commented on src/core/meshcop/tcat_agent.hpp:
```diff
@@ -83,11 +83,11 @@ class TcatAgent : public InstanceLocator, private NonCopyable
+        kGeneral         = OT_TCAT_COMMAND_CLASS_GENERAL,         ///< TCAT commands related to general operations
+        kCommissioning   = OT_TCAT_COMMAND_CLASS_COMMISSIONING,   ///< TCAT commands related to commissioning
+        kExtraction      = OT_TCAT_COMMAND_CLASS_EXTRACTION,      ///< TCAT commands related to key extraction
+        kDecommissioning = OT_TCAT_COMMAND_CLASS_DECOMMISSIONING, ///< TCAT commands related to de-commissioning
```
```suggestion
        kDecommissioning = OT_TCAT_COMMAND_CLASS_DECOMMISSIONING, ///< TCAT commands related to decommissioning
```
